src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1658,13 +1670,22 @@ public SolutionState WithFrozenPartialCompilationIncludingSpecificDocument(Docum
                         return currentPartialSolution!;
                     }
 
-                    // if we don't have one or it is stale, create a new partial solution
-                    var tracker = this.GetCompilationTracker(documentId.ProjectId);
-                    var newTracker = tracker.FreezePartialStateWithTree(this, doc, tree, cancellationToken);
+                    var newIdToProjectStateMap = _projectIdToProjectStateMap;
+                    var newIdToTrackerMap = _projectIdToTrackerMap;
 
-                    Contract.ThrowIfFalse(_projectIdToProjectStateMap.ContainsKey(documentId.ProjectId));
-                    var newIdToProjectStateMap = _projectIdToProjectStateMap.SetItem(documentId.ProjectId, newTracker.ProjectState);
-                    var newIdToTrackerMap = _projectIdToTrackerMap.SetItem(documentId.ProjectId, newTracker);
+                    for (var i = 0; i < docBuilder.Count; i++)
+                    {
+                        var doc = docBuilder[i];
+                        var tree = treeBuilder[i];
+
+                        // if we don't have one or it is stale, create a new partial solution
+                        var tracker = this.GetCompilationTracker(doc.Id.ProjectId);
+                        var newTracker = tracker.FreezePartialStateWithTree(this, doc, tree, cancellationToken);
+
+                        Contract.ThrowIfFalse(_projectIdToProjectStateMap.ContainsKey(doc.Id.ProjectId));
```

Comment:
```suggestion
                        Contract.ThrowIfFalse(newIdToProjectStateMap.ContainsKey(doc.Id.ProjectId));
```

Although the code is still technically correct, it's generally easier to scan the method and just see absolutely nothing is using the fields.